File: SOLID/code-examples/chapter-16.cpp
// FILE: chapter-16.cpp
// Chapter 16 — Refactoring — C++

#include <iostream>
#include <string>
#include <unordered_map>
#include <functional>
using namespace std;

// ❌ BEFORE Refactoring - Poor code structure
class BadCalculator {
public:
    double calc(const string& op, double a, double b) {
        if (op == "add" || op == "ADD" || op == "+") {
            return a + b;
        } else if (op == "sub" || op == "SUB" || op == "-") {
            return a - b;
        } else if (op == "mul" || op == "MUL" || op == "*") {
            return a * b;
        } else if (op == "div" || op == "DIV" || op == "/") {
            if (b == 0) {
                cout << "Error: Division by zero!" << endl;
                return 0;
            }
            return a / b;
        } else {
            cout << "Unknown operation: " << op << endl;
            return 0;
        }
    }
};

// ✅ AFTER Refactoring - Clean, maintainable code
enum class Operation {
    ADD, SUBTRACT, MULTIPLY, DIVIDE
};

struct CalculationResult {
    double value;
    bool isSuccess;
    string errorMessage;
    
    static CalculationResult success(double val) {
        return {val, true, ""};
    }
    
    static CalculationResult error(const string& msg) {
        return {0, false, msg};
    }
};

class Calculator {
private:
    unordered_map<Operation, function<CalculationResult(double, double)>> operations;
    
public:
    Calculator() {
        operations[Operation::ADD] = [](double a, double b) {
            return CalculationResult::success(a + b);
        };
        operations[Operation::SUBTRACT] = [](double a, double b) {
            return CalculationResult::success(a - b);
        };
        operations[Operation::MULTIPLY] = [](double a, double b) {
            return CalculationResult::success(a * b);
        };
        operations[Operation::DIVIDE] = [](double a, double b) {
            return b != 0 
                ? CalculationResult::success(a / b)
                : CalculationResult::error("Division by zero");
        };
    }
    
    CalculationResult calculate(Operation operation, double a, double b) {
        return operations[operation](a, b);
    }
};

int main() {
    cout << "🔧 Refactoring Example (C++)" << endl;
    cout << "============================" << endl << endl;
    
    // Before refactoring
    cout << "❌ Before refactoring:" << endl;
    BadCalculator badCalc;
    cout << "5 + 3 = " << badCalc.calc("add", 5, 3) << endl;
    cout << "10 / 0 = " << badCalc.calc("div", 10, 0) << endl;
    
    // After refactoring
    cout << endl << "✅ After refactoring:" << endl;
    Calculator goodCalc;
    
    auto result1 = goodCalc.calculate(Operation::ADD, 5, 3);
    cout << "5 + 3 = " << (result1.isSuccess ? to_string(result1.value) : result1.errorMessage) << endl;
    
    auto result2 = goodCalc.calculate(Operation::DIVIDE, 10, 0);
    cout << "10 / 0 = " << (result2.isSuccess ? to_string(result2.value) : result2.errorMessage) << endl;
    
    cout << endl << "💡 Refactoring Benefits:" << endl;
    cout << "   ✓ Better error handling" << endl;
    cout << "   ✓ Type-safe operations" << endl;
    cout << "   ✓ Easier to extend" << endl;
    cout << "   ✓ More testable" << endl;
    
    return 0;
}

File: SOLID/code-examples/chapter-16.cs
// FILE: chapter-16.cs
// Chapter 16 — Refactoring — C#

using System;
using System.Collections.Generic;

// ❌ BEFORE Refactoring - Poor code structure
class BadCalculator
{
    public double calc(string op, double a, double b)
    {
        if (op == "add" || op == "ADD" || op == "+")
        {
            return a + b;
        }
        else if (op == "sub" || op == "SUB" || op == "-")
        {
            return a - b;
        }
        else if (op == "mul" || op == "MUL" || op == "*")
        {
            return a * b;
        }
        else if (op == "div" || op == "DIV" || op == "/")
        {
            if (b == 0) 
            {
                Console.WriteLine("Error: Division by zero!");
                return 0;
            }
            return a / b;
        }
        else
        {
            Console.WriteLine("Unknown operation: " + op);
            return 0;
        }
    }
}

// ✅ AFTER Refactoring - Clean, maintainable code
public enum Operation
{
    Add, Subtract, Multiply, Divide
}

public class CalculationResult
{
    public double Value { get; }
    public bool IsSuccess { get; }
    public string ErrorMessage { get; }
    
    private CalculationResult(double value, bool isSuccess, string errorMessage)
    {
        Value = value;
        IsSuccess = isSuccess;
        ErrorMessage = errorMessage;
    }
    
    public static CalculationResult Success(double value) => new CalculationResult(value, true, null);
    public static CalculationResult Error(string message) => new CalculationResult(0, false, message);
}

public class Calculator
{
    private readonly Dictionary<Operation, Func<double, double, CalculationResult>> operations;
    
    public Calculator()
    {
        operations = new Dictionary<Operation, Func<double, double, CalculationResult>>
        {
            [Operation.Add] = (a, b) => CalculationResult.Success(a + b),
            [Operation.Subtract] = (a, b) => CalculationResult.Success(a - b),
            [Operation.Multiply] = (a, b) => CalculationResult.Success(a * b),
            [Operation.Divide] = (a, b) => b != 0 
                ? CalculationResult.Success(a / b) 
                : CalculationResult.Error("Division by zero")
        };
    }
    
    public CalculationResult Calculate(Operation operation, double a, double b)
    {
        return operations[operation](a, b);
    }
}

class Program
{
    static void Main()
    {
        Console.WriteLine("🔧 Refactoring Example (C#)");
        Console.WriteLine("==========================\n");
        
        // Before refactoring
        Console.WriteLine("❌ Before refactoring:");
        var badCalc = new BadCalculator();
        Console.WriteLine($"5 + 3 = {badCalc.calc("add", 5, 3)}");
        Console.WriteLine($"10 / 0 = {badCalc.calc("div", 10, 0)}");
        
        // After refactoring
        Console.WriteLine("\n✅ After refactoring:");
        var goodCalc = new Calculator();
        
        var result1 = goodCalc.Calculate(Operation.Add, 5, 3);
        Console.WriteLine($"5 + 3 = {(result1.IsSuccess ? result1.Value.ToString() : result1.ErrorMessage)}");
        
        var result2 = goodCalc.Calculate(Operation.Divide, 10, 0);
        Console.WriteLine($"10 / 0 = {(result2.IsSuccess ? result2.Value.ToString() : result2.ErrorMessage)}");
        
        Console.WriteLine("\n💡 Refactoring Benefits:");
        Console.WriteLine("   ✓ Better error handling");
        Console.WriteLine("   ✓ Type-safe operations");
        Console.WriteLine("   ✓ Easier to extend");
        Console.WriteLine("   ✓ More testable");
    }
}